### ISSUE TITLE
readthedocs: Broken layout on 404 page

### ISSUE THREAD

**Issue by mootari** [drud/ddev#2939]
**Describe the bug**
When accessing a stale or otherwise invalid documentation link, the layout of the returned page is broken. The reason seems to be invalid asset paths which are missing the base path. Example:
- valid page: `https://ddev.readthedocs.io/en/stable/assets/stylesheets/main.b8ac9624.min.css`
- 404 page: `https://ddev.readthedocs.io/assets/stylesheets/main.b8ac9624.min.css`

**To Reproduce**
Visit the URL https://ddev.readthedocs.io/en/latest/users/providers/user-defined/ (referenced in https://github.com/drud/ddev/releases/tag/v1.17.0-alpha5).

**Screenshots**
![image](https://user-images.githubusercontent.com/322014/114185046-8fd7a680-9945-11eb-9277-be0a1dfaf7ff.png)

**Additional context**
Ideally, stale paths should be redirected to the current paths (see e.g. https://docs.readthedocs.io/en/stable/user-defined-redirects.html).

**Comment by rfay** [comment]
This was not fixed in #2948; I note that it's *not* a problem with mkdocs. `mkdocs serve` just shows a 404 when you give it a bad URL. /cc @unn

**Comment by rfay** [comment]
@mootari, @unn if you have any ideas on how to fix this I'd love to have it work.

**Comment by mootari** [comment]
Sorry for the delay, finally found the time to look into it. Here's what I think the core problem is:

`mkdocs-material` generates relative asset paths for all pages, except for the 404 page. This page is only generated once and served at arbitrary URIs, thus using absolute asset paths. For this to work, `site_uri` must be configured correctly.

Possible solutions:
- **Redirect:** Via the readthedocs dashboard, create a [prefix redirect](https://docs.readthedocs.io/en/stable/user-defined-redirects.html#prefix-redirects) from `/assets/` to `/en/stable/assets`. This will only work properly (if at all) for the docs at `/en/stable/`. If there are other versions, these might fail with this solution.
- **Environment variables:** Read The Docs provides a few [environment variables](https://docs.readthedocs.io/en/stable/builds.html#build-environment) which can be referenced during a build. mkdocs [supports env vars](https://github.com/mkdocs/mkdocs/issues/1954) within a config yaml, but offers not means to concatenate multiple variables.  
  readthedocs does not yet expose the base URL as a single env var. This feature is tracked in https://github.com/mkdocs/mkdocs/issues/1954. Once this change is in, setting `site_uri` should become trivial.

**Comment by rfay** [comment]
Thanks so much @mootari ! 

I tried this, but it didn't seem to have any particular effect on this problem:

<img width="643" alt="Edit_Redirects___Read_the_Docs" src="https://user-images.githubusercontent.com/112444/124373570-8dd56900-dc50-11eb-9ba1-6c66070fdb98.png">

**Comment by mootari** [comment]
I see, looks like prefix redirects are the wrong choice here. Can you try an Exact Redirect from `/assets/$rest` to `/en/stable/assets/`?

**Comment by rfay** [comment]
Neither of these eemed to get me anywhere:

<img width="664" alt="Edit_Redirects___Read_the_Docs" src="https://user-images.githubusercontent.com/112444/124394720-e42cc100-dcbd-11eb-9651-64948c29859f.png">

<img width="691" alt="Edit_Redirects___Read_the_Docs" src="https://user-images.githubusercontent.com/112444/124394727-eb53cf00-dcbd-11eb-8601-a025e6c45c54.png">

**Comment by mootari** [comment]
You left out `$rest`.

**Comment by mootari** [comment]
We need a few more redirects for /css and /js. In summary, these Exact Redirects are required:
- `/assets/$rest` -> `/en/stable/assets/`
- `/css/$rest` -> `/en/stable/css/`
- `/js/$rest` -> `/en/stable/js/`

Testing can be done (more reliably than in the browser) by checking the response headers, e.g. with httpie:
```
http -h https://ddev.readthedocs.io/css/timeago.css
```

**Comment by rfay** [comment]
Looks like you've got it!

<img width="680" alt="Edit_Redirects___Read_the_Docs" src="https://user-images.githubusercontent.com/112444/124401477-05ee6e00-dce7-11eb-97a4-2e3fe22d280d.png">

location: https://ddev.readthedocs.io/en/stable/css/timeago.css

and...

<img width="749" alt="DDEV-Local_Documentation" src="https://user-images.githubusercontent.com/112444/124401534-3a622a00-dce7-11eb-8ae6-5c935869bbf9.png">


Thank you! Please poke around to see if you think it's OK.

**Comment by rfay** [comment]
The remaining problem is that once you've landed on a 404 the menus are all broken. 

For example, in the left-hand menu on a 404 page, "Using the CLI" is https://ddev.readthedocs.io/users/cli-usage/, which is itself a 404.  So once you land on a 404, you just keep getting in deeper until you to to the top. 

Seems like this is a different problem we're chasing now? But it's worse the the original in some ways because there's no way out for a casual user.

**Comment by mootari** [comment]
Yeah, I'm still seeing some 404s. So most important can likely be fixed with:
- `/readthedocs-data.js` -> `/en/stable/readthedocs-data.js`

This workaround is getting increasingly ugly, so hopefully the new readthedocs env var will be available in the new feature and allow us to properly set site_url on build.

**Comment by rfay** [comment]
Added those, thanks so much. The problem still remains that when you hit a 404 all the menus on the left side have broken links. For example "Using the CLI" points to https://ddev.readthedocs.io/users/cli-usage/, which is a 404. 

<img width="633" alt="Edit_Redirects___Read_the_Docs" src="https://user-images.githubusercontent.com/112444/124416741-e2dab300-dd14-11eb-9179-1092f1a82d4b.png">

**Comment by rfay** [comment]
@mootari wondering if they've made any progress with this upstream. Also would love to have you take a look at https://github.com/drud/ddev/issues/3232 where there's a new python-dependency problem that just cropped up with no change on our end.

**Comment by mootari** [comment]
@rfay I follow the related issues, and afair there wasn't any movement. But since I've already set up a readthedocs version of the docs for #3232 I might take a stab at this myself in the coming days.

**Comment by mootari** [comment]
We may be able to use [readthedocs-custom-steps](https://github.com/NiklasRosenstein/readthedocs-custom-steps), as described in https://github.com/readthedocs/readthedocs.org/issues/6662#issuecomment-653675848. However I'd first wait a few days for a response in https://github.com/readthedocs/readthedocs.org/issues/8260.

**Comment by rfay** [comment]
@mootari at least we get a 404 these days, your original example, https://ddev.readthedocs.io/en/latest/users/providers/user-defined/ now gets
<img width="755" alt="Cursor_and_DDEV-Local_Documentation" src="https://user-images.githubusercontent.com/112444/157764122-5fbb5210-dd7b-4568-ae97-bbbe039bf0fc.png">


It *would* be nice if it redirected, but this seems better right? Should we close this?

**Comment by mootari** [comment]
@rfay I guess? The layout problem has been sufficiently mitigated, and content redirects should probably be tracked a separate issue.

Btw, some of the related RTD issues have become more active again in the past days, so there's hope that a proper fix will be possible at some point.

**Comment by rfay** [comment]
Working on this again! All broken again due to content moves, but I think this issue probably explains how to fix it all.

**Comment by rfay** [comment]
I think this is working OK these days.
<img width="980" alt="DDEV_Docs" src="https://user-images.githubusercontent.com/112444/187007134-b4f3c996-bda3-45f4-b9e6-80fc28a4452c.png">
